sql/files/defaultdata/compare/compare.cc

```diff
@@ -72,33 +73,65 @@ FILE *openfeedback(const char *feedbackdir, const char *feedback, const char *wh
 	return res;
 }
 
-/* Test two numbers for equality, accounting for +/-INF, NaN and
- * precision. Float f2 is considered the reference value for relative
- * error.
- */
-int equal(flt f1, flt f2, flt float_abs_tol, flt float_rel_tol)
+// The behavior of std::tolower on (signed) char is undefined.
+char tolower_char(char c)
+{
+    return static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
+}
+
+bool equal_case_insensitive(std::string a, std::string b)
 {
-	flt absdiff, reldiff;
+	for (char &c : a) c = tolower_char(c);
+	for (char &c : b) c = tolower_char(c);
+
+	return a==b;
+}
+
+/* Test two floating-point numbers for equality, accounting for +/-INF, NaN, and precision.
+ * Float `jval` is considered the reference value for relative error.
+ */
+void compare_float(std::string judge, std::string team, flt jval, flt tval, flt float_abs_tol, flt float_rel_tol, std::string extra_msg) {
 	/* Finite values are compared with some tolerance */
-	if ( std::isfinite(f1) && std::isfinite(f2) ) {
-		absdiff = fabsl(f1-f2);
-		reldiff = fabsl((f1-f2)/f2);
-		return !(absdiff > float_abs_tol && reldiff > float_rel_tol);
-	}
+	if (std::isfinite(tval) && std::isfinite(jval)) {
+		flt absdiff = fabsl(tval-jval);
+		flt reldiff = fabsl((tval-jval)/jval);
+		if (float_abs_tol >= 0 && float_rel_tol >= 0) {
+			if (absdiff > float_abs_tol && reldiff > float_rel_tol) {
+				wrong_answer("Too large difference.\n Judge: %s\n Team: %s\n Absolute difference: %Lg (tolerance: %Lg%s)\n Relative difference: %Lg (tolerance: %Lg%s)%s",
+				             judge.c_str(), team.c_str(),
+				             absdiff, float_abs_tol, absdiff > float_abs_tol ? ", exceeded" : "",
+				             reldiff, float_rel_tol, reldiff > float_rel_tol ? ", exceeded" : "",
```

Comment:
I'm gonna be very nit-picky and refactor these conditionals now. Looking at it again, the conditions are always `true` :grin: 
```suggestion
				wrong_answer("Too large difference.\n Judge: %s\n Team: %s\n Absolute difference: %Lg (tolerance: %Lg, exceeded)\n Relative difference: %Lg (tolerance: %Lg, exceeded)%s",
				             judge.c_str(), team.c_str(),
				             absdiff, float_abs_tol, reldiff, float_rel_tol,
```

I tested this on the BAPC 2022 problems again (locally), nothing breaks :slightly_smiling_face: